functions/post_data.go

```diff
@@ -0,0 +1,147 @@
+package functions
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const timeoutSeconds = 10 // the max number of seconds to wait for a response
+var errURLMissing = errors.New("error calling PostData; URL is required")
+var errURLEmpty = errors.New("error calling PostData; URL is empty")
+
+// PostData sends an HTTP POST request to a given URL with the data provided,
+// returning either the whole response body or a specified JSON element.
+func PostData(existingHeaders map[string]string, args []string) (string, error) {
+	// Ensure that we have at least one argument for the URL
+	if len(args) == 0 {
+		return "", errURLMissing
+	}
+
+	// Get the URL, response element, and request body from the args
+	url, responseElement, err := argsToURLResponseElement(args)
+	if err != nil {
+		return "", err
+	}
+	var requestBody string
+	if len(args) > 2 {
+		requestBody = argsToRequestBody(existingHeaders, args[2:])
+	}
+
+	// Send the request
+	request, err := http.NewRequest("POST", url, bytes.NewBuffer([]byte(requestBody)))
+	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	client := &http.Client{}
+	client.Timeout = time.Second * timeoutSeconds
+	response, err := client.Do(request)
+	if err != nil {
+		return "", err
+	}
+	responseBody, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return "", err
+	}
+	defer response.Body.Close()
+
+	// Get the specific element requested from the response
+	var result string
+	jsonBody := make(map[string]json.RawMessage)
+	err = json.Unmarshal(responseBody, &jsonBody)
+	if err != nil {
+		// This must be an array, so just return the whole response since we can't determine which element to return
+		responseElement = ""
+	}
+	if responseElement != "" {
+		// Return the value associated with the key responseElement,
+		// either a plain string for primitives, or a JSON string for objects/arrays
+		value := jsonBody[responseElement]
```

Comment:
Here I would use a check like this: `value, ok := jsonBody[responseElement]`

Also, what if you want to access a value deeper in the response? If you allow specifying the access using dot notation `form.field.value`, you can use recursive `json.Unmarshal(body, json.RawMessage)` to retrieve each element: https://play.golang.org/p/EqgkqDGGna3

Reply:
I like the dot notation idea, and I've mostly implemented what you've suggested, however there is one case where I still need to differentiate between object/array vs. primitive, and that's when I convert from the raw JSON to a byte slice (basically the last step before we return an element). This handles things like returning strings without unnecessary quotes around them and returning objects/arrays in more standardized from (insignificant space characters removed).

Reply:
Mm, interesting. What do you think about leveraging a library to do this bit?
This brings us close to something like jq or jsonpath and you can delegate the search to it, like https://gjson.dev/ (which is neither of those but seems well tested).